File: app/src/main/java/com/botijasoftware/utils/Vector3i.java
package com.botijasoftware.utils;

public class Vector3i {
	
	public Vector3i() {
		X = Y = Z = 0;
	}

	public Vector3i(int val) {
		X = val;
		Y = val;
		Z = val;
	}
	public Vector3i(int valx, int valy, int valz) {
		X = valx;
		Y = valy;
		Z = valz;
	}
	
	public void setValue(int val) {
		X = val;
		Y = val;
		Z = val;
	}
	
	public void setValue(int valx, int valy, int valz) {
		X = valx;
		Y = valy;
		Z = valz;
	}
	
	public void setValue(Vector3i v) {
		X = v.X;
		Y = v.Y;
		Z = v.Z;
	}
	
	public Vector3i clone() {
		return new Vector3i(X,Y,Z);
	}
	
	public Vector3i normalize() {
		float len = length();
		if (len != 0.0f) {
			X = (int) (X / len);
			Y = (int) (Y / len);
			Z = (int) (Z / len);
		}
		return this;
	}
	
	public Vector3i add(Vector3i v) {
		X += v.X;
		Y += v.Y;
		Z += v.Z;
		return this;
	}
	
	public Vector3i add(int val) {
		X += val;
		Y += val;
		Z += val;
		return this;
	}
	
	public Vector3i sub(Vector3i v) {
		X -= v.X;
		Y -= v.Y;
		Z -= v.Z;
		return this;
	}
	
	public Vector3i sub(int val) {
		X -= val;
		Y -= val;
		Z -= val;
		return this;
	}
	
	public Vector3i mul(Vector3i v) {
		X *= v.X;
		Y *= v.Y;
		Z *= v.Z;
		return this;
	}
	
	public Vector3i mul(int val) {
		X *= val;
		Y *= val;
		Z *= val;
		return this;
	}
	
	public Vector3i div(Vector3i v) {
		if (v.X != 0.0f && v.Y != 0.0f && v.Z != 0.0f){
			X /= (X / v.X);
			Y /= (Y / v.X);
			Z /= (Z / v.X);
		}
		return this;
	}
	
	public Vector3i div(int val) {
		if (val != 0.0f){
			float inv = 1.0f/val;
			X = (int)(X * inv);
			Y = (int)(Y * inv);
			Z = (int)(Z * inv);
		}
		return this;
	}
	
	public float lengthsq() {
		return X * X + Y * Y + Z * Z;
	}
	
	public float length() {
		return (float)Math.sqrt( X * X + Y * Y + Z * Z);
	}
	
	public boolean isZero() {
		return (X == 0 && Y == 0 && Z == 0);
	}
	
	public boolean equals(int val) {
		return (X == val && Y == val && Z == val);
	}
	
	public boolean equals(Vector3i v) {
		return (X == v.X && Y == v.Y && Z == v.Z);
	}
	
	public float distancesq(Vector3i v) {
		return ((X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y) + (Z - v.Z) * (Z - v.Z));
	}
	
	public float distance(Vector3i v) {
		return (float)Math.sqrt( (X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y) + (Z - v.Z) * (Z - v.Z) );
	}
	
	public float dot(Vector3i v) {
		return X * v.X + Y * v.Y + Z * v.Z;
	}
	
	public Vector3i cross(Vector3i v) {
		float nx = Y * v.Z - Z * v.Y;
		float ny = Z * v.X - X * v.Z;
		float nz = X * v.Y - Y * v.X;
		X = (int)nx;
		Y = (int)ny;
		Z = (int)nz;
		return this;
	}
	
	public Vector3i reflect(Vector3i v) {
		//i - (2 * n * dot(i, n))
		sub(v.mul( (int)(2 * dot(v))));
		return this;
	}
	

	public Vector3i zero() {
		X = 0;
		Y = 0;
		Z = 0;
		return this;
	}
	
	public Vector3i lerp(Vector3i v, float factor) {
		if (factor <=0.0f) {
			return this;
		}
		else if (factor >= 1.0f) {
			X = v.X;
			Y = v.Y;
			Z = v.Z;
			return this;
		}
		
		X = X + (int)(factor * ( v.X - X));
		Y = Y + (int)(factor * ( v.Y - Y));
		Z = Z + (int)(factor * ( v.Z - Z));
		
		return this;
	}
	
	public static Vector3i cross(Vector3i v, Vector3i u) {
		int nx = v.Y * u.Z - v.Z * u.Y;
		int ny = v.Z * u.X - v.X * u.Z;
		int nz = v.X * u.Y - v.Y * u.X;		
		return new Vector3i(nx, ny, nz);
	}
	
	public static Vector3i orthoNormalize(Vector3i n, Vector3i u) {
		n.normalize();
		Vector3i v = cross(n,u);
		v.normalize();
		return cross(v,n);		
	}
	
	public static Vector3i lerp(Vector3i v, Vector3i u, float factor) {
		if (factor <=0.0f) {
			return v.clone();
		}
		else if (factor >= 1.0f) {
			return u.clone();
		}
		
		int x = v.X + (int)(factor * ( u.X - v.X));
		int y = v.Y + (int)(factor * ( u.Y - v.Y));
		int z = v.Z + (int)(factor * ( u.Z - v.Z));
		
		return new Vector3i(x,y,z);
	}
	
	public int X;
	public int Y;
	public int Z;
	public static final Vector3i ZERO = new Vector3i(0, 0, 0);
    public static final Vector3i ONE = new Vector3i(1, 1, 1);
    public static final Vector3i RIGHT = new Vector3i(1, 0, 0);
    public static final Vector3i UP = new Vector3i(0, 1, 0);
    public static final Vector3i FORWARD = new Vector3i(0, 0, 1);

}

File: app/src/main/java/com/botijasoftware/utils/Vector3i.kt
package com.botijasoftware.utils

class Vector3i {

    constructor() {
        Z = 0
        Y = Z
        X = Y
    }

    constructor(value: Int) {
        X = value
        Y = value
        Z = value
    }

    constructor(valx: Int, valy: Int, valz: Int) {
        X = valx
        Y = valy
        Z = valz
    }

    fun setValue(value: Int) {
        X = value
        Y = value
        Z = value
    }

    fun setValue(valx: Int, valy: Int, valz: Int) {
        X = valx
        Y = valy
        Z = valz
    }

    fun setValue(v: Vector3i) {
        X = v.X
        Y = v.Y
        Z = v.Z
    }

    fun clone(): Vector3i {
        return Vector3i(X, Y, Z)
    }

    fun normalize(): Vector3i {
        val len = length()
        if (len != 0.0f) {
            X = (X / len).toInt()
            Y = (Y / len).toInt()
            Z = (Z / len).toInt()
        }
        return this
    }

    fun add(v: Vector3i): Vector3i {
        X += v.X
        Y += v.Y
        Z += v.Z
        return this
    }

    fun add(value: Int): Vector3i {
        X += value
        Y += value
        Z += value
        return this
    }

    fun sub(v: Vector3i): Vector3i {
        X -= v.X
        Y -= v.Y
        Z -= v.Z
        return this
    }

    fun sub(value: Int): Vector3i {
        X -= value
        Y -= value
        Z -= value
        return this
    }

    fun mul(v: Vector3i): Vector3i {
        X *= v.X
        Y *= v.Y
        Z *= v.Z
        return this
    }

    fun mul(value: Int): Vector3i {
        X *= value
        Y *= value
        Z *= value
        return this
    }

    operator fun div(v: Vector3i): Vector3i {
        if (v.X != 0 && v.Y != 0 && v.Z != 0) {
            X /= X / v.X
            Y /= Y / v.X
            Z /= Z / v.X
        }
        return this
    }

    operator fun div(value: Int): Vector3i {
        if (value != 0) {
            val inv = 1.0f / value
            X = (X * inv).toInt()
            Y = (Y * inv).toInt()
            Z = (Z * inv).toInt()
        }
        return this
    }

    fun lengthsq(): Float {
        return (X * X + Y * Y + Z * Z).toFloat()
    }

    fun length(): Float {
        return Math.sqrt((X * X + Y * Y + Z * Z).toDouble()).toFloat()
    }

    val isZero: Boolean
        get() = X == 0 && Y == 0 && Z == 0

    fun equals(value: Int): Boolean {
        return X == value && Y == value && Z == value
    }

    fun equals(v: Vector3i): Boolean {
        return X == v.X && Y == v.Y && Z == v.Z
    }

    fun distancesq(v: Vector3i): Float {
        return ((X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y) + (Z - v.Z) * (Z - v.Z)).toFloat()
    }

    fun distance(v: Vector3i): Float {
        return Math.sqrt(((X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y) + (Z - v.Z) * (Z - v.Z)).toDouble()).toFloat()
    }

    fun dot(v: Vector3i): Float {
        return (X * v.X + Y * v.Y + Z * v.Z).toFloat()
    }

    fun cross(v: Vector3i): Vector3i {
        val nx = (Y * v.Z - Z * v.Y)
        val ny = (Z * v.X - X * v.Z)
        val nz = (X * v.Y - Y * v.X)
        X = nx
        Y = ny
        Z = nz
        return this
    }

    fun reflect(v: Vector3i): Vector3i {
        //i - (2 * n * dot(i, n))
        sub(v.mul((2 * dot(v)).toInt()))
        return this
    }


    fun zero(): Vector3i {
        X = 0
        Y = 0
        Z = 0
        return this
    }

    fun lerp(v: Vector3i, factor: Float): Vector3i {
        if (factor <= 0.0f) {
            return this
        } else if (factor >= 1.0f) {
            X = v.X
            Y = v.Y
            Z = v.Z
            return this
        }

        X += (factor * (v.X - X)).toInt()
        Y += (factor * (v.Y - Y)).toInt()
        Z += (factor * (v.Z - Z)).toInt()

        return this
    }

    var X: Int = 0
    var Y: Int = 0
    var Z: Int = 0

    companion object {

        fun cross(v: Vector3i, u: Vector3i): Vector3i {
            val nx = v.Y * u.Z - v.Z * u.Y
            val ny = v.Z * u.X - v.X * u.Z
            val nz = v.X * u.Y - v.Y * u.X
            return Vector3i(nx, ny, nz)
        }

        fun orthoNormalize(n: Vector3i, u: Vector3i): Vector3i {
            n.normalize()
            val v = cross(n, u)
            v.normalize()
            return cross(v, n)
        }

        fun lerp(v: Vector3i, u: Vector3i, factor: Float): Vector3i {
            if (factor <= 0.0f) {
                return v.clone()
            } else if (factor >= 1.0f) {
                return u.clone()
            }

            val x = v.X + (factor * (u.X - v.X)).toInt()
            val y = v.Y + (factor * (u.Y - v.Y)).toInt()
            val z = v.Z + (factor * (u.Z - v.Z)).toInt()

            return Vector3i(x, y, z)
        }

        val ZERO = Vector3i(0, 0, 0)
        val ONE = Vector3i(1, 1, 1)
        val RIGHT = Vector3i(1, 0, 0)
        val UP = Vector3i(0, 1, 0)
        val FORWARD = Vector3i(0, 0, 1)
    }

}
